### ISSUE TITLE
allow dynamic assignment of type

### ISSUE THREAD

**Issue by andreaseger** [jsonapi-rb/jsonapi-serializable#53]
with polymorphic types which can be serialized with the same serializable resource class it would be great if the type could be set dynamically i.e. like this

```ruby
class SerializableSample < JSONAPI::Serializable::Resource
  type do
    @object.type
  end
  attribute :created_at
end
```

that way the concrete type could be set in the response instead of using the polymorphic type

**Comment by beauby** [comment]
I'm interested in your use-case. If all your resources have the same representation, does the `type` really matter to the client?

As a workaround for now, you can override the `jsonapi_type` method as follows:
```ruby
class SerializableSample < JSONAPI::Serializable::Resource
  def jsonapi_type
    @object.type
  end
  attribute :created_at
end
```

**Comment by andreaseger** [comment]
We have some polymorphic resources which in some cases will be serialized only with the common attributes between all sub-types. Although in theory it would be sufficient to simply send the clients the polymorphic type we'd like to still give them the information regarding which concrete subtype a single resource entity is. This gives the clients to possibility to for example display the concrete subtype in a different way or request for certain sub-types more information.

I guess your workaround will work for now and is better than what we where using right now.

**Comment by nilsding** [comment]
@beauby This workaround does sadly not work, because the `#as_jsonapi` method for the type uses the `@_type` ivar instead of `jsonapi_type`: [/lib/jsonapi/serializable/resource/type.rb#L24](https://github.com/jsonapi-rb/jsonapi-serializable/blob/2c01d15b50d3dbeca6c7cbcf62a672d3b0836aea/lib/jsonapi/serializable/resource/type.rb#L24)
The same is true with `id` and other values, but these can be set using a block.

**Comment by beauby** [comment]
@nilsding You're right, sorry. Accepting a block for `type` is a trivial change to make (mainly copy what's being done for `id`) in the codebase. I removed it at some point because I couldn't find a suitable use-case for it, but I'll re-add it when I have a bit more time. I'll gladly accept a PR if someone wants to get it in quick.

**Comment by beauby** [comment]
In the meantime, a possible workaround is:
```ruby
class PolymorphicSerializable < JSONAPI::Serializable::Resource
  ...
end

["Foo", "Bar", "Baz"].each do |k|
  Object.const_set("Serializable#{k}", Class.new(PolymorphicSerializable))
end
```

**Comment by beauby** [comment]
Closed by #54.
